lib/core/src/Cardano/Wallet.hs

```diff
@@ -559,37 +562,30 @@ restoreBlocks ctx wid blocks nodeTip = do
                     then Ready
                     else Restoring progress'
 
-        -- NOTE:
-        -- We cast `k` and `nodeHeight` to 'Integer' since at a given point
-        -- in time, `k` may be greater than the tip.
-        let (Quantity k) = bp ^. #getEpochStability
-        let (Quantity nodeHeight) = nodeTip ^. #blockHeight
-        let bhUnstable :: Integer
-            bhUnstable = fromIntegral nodeHeight - fromIntegral k
-
-        forM_ (NE.init cps) $ \cp -> do
-            let (Quantity bh) = blockHeight $ currentTip cp
-            when (fromIntegral bh >= bhUnstable) $
+        let makeCheckpoint :: Wallet s t -> ExceptT ErrNoSuchWallet IO ()
+            makeCheckpoint cp = do
+                liftIO $ logInfo tr $
+                    "Creating checkpoint at " <> pretty (currentTip cp)
                 DB.putCheckpoint db (PrimaryKey wid) cp
 
-        -- NOTE:
-        -- Always store the last checkpoint from the batch and all new
-        -- transactions.
-        let cpLast = NE.last cps
-        let Quantity bhLast = blockHeight $ currentTip cpLast
-        let meta' = calculateMetadata (view #slotId $ currentTip cpLast)
-        DB.putCheckpoint db (PrimaryKey wid) cpLast
+        let unstable = sparseCheckpoints k nodeTip
+        forM_  cps $ \cp -> do
```

Comment:
hypothesis : this nonlinearity when we deal with additional epoch comes from this `cps`